velox/vector/tests/LazyVectorTest.cpp

```diff
@@ -223,7 +255,52 @@ TEST_F(LazyVectorTest, nestedLazyRowsInDictionary) {
   assertEqualVectors(child, lazyInnerRowVector);
 }
 
-TEST_F(LazyVectorTest, nestedLazyRowsPartiallyLoaded) {
+TEST_F(
+    LazyVectorTest,
+    dictionaryOverLazyRowVectorWithDictionaryOverLazyChildren) {
+  // Input: dict(row(lazy(dict(row(lazy)))))
+  constexpr vector_size_t size = 1000;
+
+  auto lazyVectorA = vectorMaker_.lazyFlatVector<int32_t>(
+      size,
+      [](vector_size_t i) { return i % 5; },
+      [](vector_size_t i) { return i % 7 == 0; });
+
+  VectorPtr innerRowVector = vectorMaker_.rowVector({lazyVectorA});
+  VectorPtr innerDict = BaseVector::wrapInDictionary(
+      nullptr,
+      makeIndices(size, [](auto row) { return row; }),
+      size,
+      innerRowVector);
+
+  VectorPtr rowVector = vectorMaker_.rowVector({std::make_shared<LazyVector>(
+      pool_.get(),
+      innerRowVector->type(),
+      size,
+      std::make_unique<SimpleVectorLoader>(
+          [&](RowSet /* rowSet */) { return innerDict; }))});
+
+  VectorPtr dict = BaseVector::wrapInDictionary(
+      nullptr,
+      makeIndices(size, [](auto row) { return row; }),
+      size,
+      rowVector);
+
+  EXPECT_TRUE(isLazyNotLoaded(*dict));
+
+  DecodedVector decodedVector;
+  SelectivityVector baseRows;
+  LazyVector::ensureLoadedRows(
+      dict, SelectivityVector(size, true), decodedVector, baseRows);
+
+  EXPECT_FALSE(isLazyNotLoaded(*dict.get()));
+
+  // Ensure encoding layer is correctly initialized after lazy loading.
+  EXPECT_NO_THROW(dict->mayHaveNullsRecursive());
+  EXPECT_NO_THROW(innerDict->mayHaveNullsRecursive());
+}
```

Comment:
nit for completion sake can you validate few values from inner lazy vector ?

Reply:
The lazy wrapper (custom loader) here just returns the vector directly without materializing anything new. This allows us to only check for state changes like, is the lazy loaded or the state has been correctly materialized.
I will therefore add a check the verifies that the loaded vector points the same VectorPtr that we expect.